python/pyarrow/src/arrow/python/arrow_to_pandas.cc

```diff
@@ -620,37 +620,42 @@ inline Status ConvertAsPyObjects(const PandasOptions& options, const ChunkedArra
   using ArrayType = typename TypeTraits<Type>::ArrayType;
   using Scalar = typename MemoizationTraits<Type>::Scalar;
 
-  ::arrow::internal::ScalarMemoTable<Scalar> memo_table(options.pool);
-  std::vector<PyObject*> unique_values;
-  int32_t memo_size = 0;
-
-  auto WrapMemoized = [&](const Scalar& value, PyObject** out_values) {
-    int32_t memo_index;
-    RETURN_NOT_OK(memo_table.GetOrInsert(value, &memo_index));
-    if (memo_index == memo_size) {
-      // New entry
-      RETURN_NOT_OK(wrap_func(value, out_values));
-      unique_values.push_back(*out_values);
-      ++memo_size;
-    } else {
-      // Duplicate entry
-      Py_INCREF(unique_values[memo_index]);
-      *out_values = unique_values[memo_index];
-    }
-    return Status::OK();
-  };
-
-  auto WrapUnmemoized = [&](const Scalar& value, PyObject** out_values) {
-    return wrap_func(value, out_values);
-  };
+  std::shared_ptr<::arrow::internal::ScalarMemoTable<Scalar>> memo_table = nullptr;
+  std::shared_ptr<std::vector<PyObject*>> unique_values = nullptr;
+  std::shared_ptr<int32_t> memo_size = std::make_shared<int32_t>(0);
+
+  std::function<Status(const typename MemoizationTraits<Type>::Scalar&, PyObject**)>
```

Comment:
`std::function` will add a layer of indirection. Did you try to measure performance to see if that matters?

Reply:
Or, better, just keep the existing `WrapMemoized` and `WrapUnmemoized`.

Reply:
I had a lot of challenges with building and/or running the program without the shared pointers. Without them, the lambda function at run-time or build-time declared things were out of scope or not declared. I can share with you what I had previously, and how it froze when I ran it. 

If I keep the `WrapMemoized` and `WrapUnmemoized`, then I would need an additional `if statement` later on, but maybe that is fine.

Reply:
Well, I can try to do something.

Reply:
Ok, it should be better now!